pkg/planner/core/rule_outer_to_inner_join.go

```diff
@@ -0,0 +1,196 @@
+	} else if p.JoinType == AntiSemiJoin {
+		innerTable = innerTable.ConvertOuterToInnerJoin(predicates)
+		outerTable = outerTable.ConvertOuterToInnerJoin(combinedCond)
+	} else {
+		innerTable = innerTable.ConvertOuterToInnerJoin(predicates)
+		outerTable = outerTable.ConvertOuterToInnerJoin(predicates)
+	}
+
+	if switchChild {
+		p.SetChild(0, outerTable)
+		p.SetChild(1, innerTable)
+	} else {
+		p.SetChild(0, innerTable)
+		p.SetChild(1, outerTable)
+	}
+
+	return p
+}
+
+func (*convertOuterToInnerJoin) name() string {
+	return "convert_outer_to_inner_joins"
+}
+
+// ConvertOuterToInnerJoin implements base.LogicalPlan ConvertOuterToInnerJoin interface.
+func (s *LogicalSelection) ConvertOuterToInnerJoin(predicates []expression.Expression) base.LogicalPlan {
+	p := s.self.(*LogicalSelection)
+	combinedCond := append(predicates, p.Conditions...)
+	child := p.Children()[0]
+	child = child.ConvertOuterToInnerJoin(combinedCond)
+	p.SetChildren(child)
+	return p
+}
+
+// ConvertOuterToInnerJoin implements base.LogicalPlan ConvertOuterToInnerJoin interface.
+func (s *LogicalProjection) ConvertOuterToInnerJoin(predicates []expression.Expression) base.LogicalPlan {
+	p := s.self.(*LogicalProjection)
+	canBePushed, _ := BreakDownPredicates(p, predicates)
+	child := p.Children()[0]
+	child = child.ConvertOuterToInnerJoin(canBePushed)
+	p.SetChildren(child)
+	return p
+}
+
+// allConstants checks if only the expression has only constants.
+func allConstants(expr expression.Expression) bool {
+	switch v := expr.(type) {
+	case *expression.ScalarFunction:
+		for _, arg := range v.GetArgs() {
+			if !allConstants(arg) {
+				return false
+			}
+		}
+		return true
+	case *expression.Constant:
+		return true
+	}
+	return false
+}
+
+// isNullFiltered takes care of complex predicates like this:
+// isNullFiltered(A OR B) = isNullFiltered(A) AND isNullFiltered(B)
+// isNullFiltered(A AND B) = isNullFiltered(A) OR isNullFiltered(B)
+func isNullFiltered(ctx base.PlanContext, innerSchema *expression.Schema, predicate expression.Expression, outerSchema *expression.Schema) bool {
+	// The expression should reference at least one field in innerSchema or all constants.
+	if !expression.ExprReferenceSchema(predicate, innerSchema) && !allConstants(predicate) {
+		return false
+	}
+
+	switch expr := predicate.(type) {
+	case *expression.ScalarFunction:
+		if expr.FuncName.L == ast.LogicAnd {
+			if isNullFiltered(ctx, innerSchema, expr.GetArgs()[0], outerSchema) {
+				return true
+			}
+			return isNullFiltered(ctx, innerSchema, expr.GetArgs()[0], outerSchema)
+		} else if expr.FuncName.L == ast.LogicOr {
+			if !(isNullFiltered(ctx, innerSchema, expr.GetArgs()[0], outerSchema)) {
+				return false
+			}
+			return isNullFiltered(ctx, innerSchema, expr.GetArgs()[1], outerSchema)
+		} else {
+			return util.IsNullRejected(ctx, innerSchema, expr)
+		}
+	default:
+		return util.IsNullRejected(ctx, innerSchema, predicate)
```

Comment:
Can be simplified to:
```suggestion
		}
	default:
		return util.IsNullRejected(ctx, innerSchema, predicate)
```